solana: testnet cli init at 1.9.2

happysalada opened the pull request:
###### Motivation for this change

I've had some troubles running the tests. I'm leaving things commented out just for when I get more time to investigate.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

bachp commented on pkgs/applications/blockchains/solana/default.nix:
```diff
@@ -0,0 +1,48 @@
+  # OPENSSL_NO_VENDOR = 1;
+  # OPENSSL_LIB_DIR = "${openssl.out}/lib";
+  # OPENSSL_DIR="${lib.getDev openssl}";
+  doCheck = false;
```
@happysalada I also had some troubles with running protobuf when building lighthouse.
See [here](https://github.com/NixOS/nixpkgs/blob/75a850f80a5d2bf5f9a331f81e089d0483c48daa/pkgs/applications/blockchains/lighthouse/default.nix#L57) for the settings I had to add.

happysalada replied:
Thank you!
I've added the two variables "PROTOC" and "PROTOC_INCLUDE", but it still fails. I'm leaving those commented out so that people can try.
The error is just "No such file or directory". No idea what is making that.

ruuda replied:
Just adding `protobuf` to the `nativeBuildInputs` was enough to make it compile in this case.

happysalada commented:
@ruuda you might be interested in this.

ruuda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -30042,6 +30042,10 @@ with pkgs;
 
   sumokoin = callPackage ../applications/blockchains/sumokoin { boost = boost165; };
 
+  solana = callPackage ../applications/blockchains/solana {
```
Should we call just the `solana` binary the “solana” package? I would expect “solana” to provide everything that the upstream installer does (`solana`, but also `solana-keygen`, `solana-validator`, `solana-test-validator`, `spl-token`, `cargo-build-bpf`, etc.)

But I suppose we can start with just the `solana` binary, and maybe even many of the others as well, and then add the remaining things once we figure out how to do that?

happysalada replied:
You're right here. I was wondering also what best to call it. Perhaps solana-cli ?

happysalada replied:
I called the package name "solana-testnet-cli". I had some hesitations about adding testnet in the name. I was wondering if perhaps we should also add the mainnet which is on a different version. I thought about starting with testnet so I could test it, then later on adding mainnet when I was confident that all is running as expected.

ruuda replied:
> I was wondering if perhaps we should also add the mainnet which is on a different version. I thought about starting with testnet so I could test it, then later on adding mainnet when I was confident that all is running as expected.

Yeah this is a bit messy, because upstream releases just have a version number and sometimes they say “recommended for mainnet” and sometimes they say “this is a testnet release not reccommended for mainnet-beta”. It looks like they follow the release train model, where currently for example 1.8.x is recommended for mainnet, and 1.9.x for testnet, and at some point 1.9.y will become recommended for mainnet. So maybe we should offer the latest patch version of the latest testnet and mainnet release side by side? `solana18` and `solana19`, just like we have `python38` and `python39`? And then we can have a `solana` alias for the mainnet one.

ruuda replied:
> You're right here. I was wondering also what best to call it. Perhaps solana-cli ?

After some thought, I think it’s fine to call it `solana`; if we add more binaries to the same package later that shouldn’t break users who were using the package for just the `solana` binary.

happysalada replied:
how about having `solana-mainnet` and `solana-testnet` then ? It seems more transparent for end users. Otherwise they have to look up versions and what they mean.

ruuda commented on pkgs/applications/blockchains/solana/default.nix:
```diff
@@ -0,0 +1,44 @@
+, AppKit
+, openssl
+, pkg-config
+, rustfmt
```
It needs `udev`, `zlib`, and also `rocksdb` for the validator, although for just the `solana` CLI, `rocksdb` may not be needed. Currently this does not build on Linux:

```
builder for '/nix/store/6bpwzj3k3864vcv3x9q559f0gzqas9q9-solana-testnet-cli-1.9.2.drv' failed with exit code 101; last 10 log lines:
    thread 'main' panicked at 'Unable to find libudev: `"pkg-config" "--libs" "--cflags" "libudev"` did not exit successfully: exit status: 1
    --- stderr
    Package libudev was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libudev.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libudev' found
    ', /build/solana-testnet-cli-1.9.2-vendor.tar.gz/hidapi/build.rs:55:54
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  warning: build failed, waiting for other jobs to finish...
  error: build failed
```

happysalada replied:
Let me add those.

I tried to run the checkphase. But on darwin, it fails with a missing framework that doesn't exist on nixos. So we won't be able to run the checkphase ever on darwin (most probably). I left the checkphase on linux to run in case someone can test. If you find out missing dependencies for the checkphase (most probably rocksdb for example) just let me know and I can add those to the checkInputs. If we find that in the end we still can't run the tests, we can always leave the checkphase to false.

ruuda replied:
I ran into some issues with the tests in https://github.com/NixOS/nixpkgs/pull/121009, but that was for the full test suite, not limited to `cli`. A few of those issues have been fixed in the meantime, but most haven’t. I can try and build this, maybe the tests for just `cli` are actually fine.

ruuda reviewed (commented):
Thanks for adding this! It needs a few more dependencies before it builds on Linux though.

happysalada commented:
I tried taking care of the first round of comments, let me know if anything.

ruuda commented:
I’m trying to build once more on Linux, but I’m running into a really puzzling issue with the vendored dependencies. I opened #153774 for that.

happysalada commented:
No worries, we can wait for a bit.
let me know what you think about having `solana-mainnet` and `solana-testnet` rather than the versions.

ruuda commented:
#### Binaries

I needed to add `profobuf` to `nativeBuildInputs` (for `protoc` which is needed to compile the `prost-build` crate). After that the binary could be built, but the tests not yet.

#### Tests

I added `llvm` and `llvmPackages.libclang.lib` and `rocksdb` to `checkInputs`, and the following `preCheck`:
```nix
  preBuild = ''
    export LLVM_CONFIG_PATH="${llvm}/bin/llvm-config";
    export LIBCLANG_PATH="${llvmPackages.libclang.lib}/lib";
    # Used by build.rs in the rocksdb-sys crate. If we don't set these, it would
    # try to build RocksDB from source.
    export ROCKSDB_INCLUDE_DIR="${rocksdb}/include";
    export ROCKSDB_LIB_DIR="${rocksdb}/lib";
  '';
```

This made some progress, but now building `rocksdb-sys` fails:

```
error: failed to run custom build command for `librocksdb-sys v6.20.3`

Caused by:
  process didn't exit successfully: `/build/source/target/release/build/librocksdb-sys-966b98292a2d40a1/build-script-build` (exit status: 101)
  --- stderr
  /nix/store/9bps2z67qs1mghyvk8pb6qxpwf8jpfby-rocksdb-6.26.1/include/rocksdb/c.h:65:10: fatal error: 'stdarg.h' file not found
  /nix/store/9bps2z67qs1mghyvk8pb6qxpwf8jpfby-rocksdb-6.26.1/include/rocksdb/c.h:65:10: fatal error: 'stdarg.h' file not found, err: true
  thread 'main' panicked at 'unable to generate rocksdb bindings: ()', /build/solana-testnet-cli-1.9.2-vendor.tar.gz/librocksdb-sys/build.rs:44:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

That header is provided in `${llvmPackages.libclang.lib}/lib/clang/${llvmPackages.libclan.version}/include`, but adding that to the include path seems like the wrong way to go, I suspect it’s not using the right compiler, but I don’t really have time to go down this rabbit hole right now. Maybe for now it’s best to just disable the tests.

#### Package name

> let me know what you think about having solana-mainnet and solana-testnet rather than the versions.

I have no strong opinion on this, I we have both numbered and named cases in Nixpkgs already (e.g. `python{38,39}`, but also `nginx{Stable,Mainline}`. In all cases we do also have one without the suffix, so I would also add a `solana` package that is an alias for `solana-mainnet`.

happysalada commented:
Ok, I think I've addressed all the concerns.

- I've renamed it to solana-testnet
- I've added the environment variables and dependencies you mentioned.
- I've checked again compilation.

I would like to merge testnet and then we can make a separate PR for mainnet. We might run into additional troubles (compilation or otherwise). Also it would be good to have it available earlier for other people to potentially test and give feedback.

Let me know if you have any other comment on testnet.

ruuda commented on pkgs/applications/blockchains/solana/default.nix:
```diff
@@ -0,0 +1,60 @@
+
+  buildAndTestSubdir = "cli";
+
+  nativeBuildInputs = lib.optionals stdenv.isLinux [ protobuf ];
+  buildInputs = lib.optionals stdenv.isLinux [ pkg-config udev zlib ] ++ lib.optionals stdenv.isDarwin [ IOKit Security AppKit ];
```
```suggestion
  nativeBuildInputs = lib.optionals stdenv.isLinux [ protobuf pkg-config ];
  buildInputs = lib.optionals stdenv.isLinux [ udev zlib ] ++ lib.optionals stdenv.isDarwin [ IOKit Security AppKit ];
```

It now fails on
```
error: failed to run custom build command for `hidapi v1.3.0`

Caused by:
  process didn't exit successfully: `/build/source/target/release/build/hidapi-aa6c9678cc52e062/build-script-build` (exit status: 101)
(...)
  thread 'main' panicked at 'Unable to find libudev: Could not run `"pkg-config" "--libs" "--cflags" "libudev"`
  The pkg-config command could not be found.

  Most likely, you need to install a pkg-config package for your OS.
  Try `apt install pkg-config`, or `yum install pkg-config`,
  or `pkg install pkg-config` depending on your distribution.

  If you've already installed it, ensure the pkg-config command is one of the
  directories in the PATH environment variable.
```

Moving `pkg-config` from `buildInputs` to `nativeBuildInputs` makes it compile.

ruuda commented on pkgs/applications/blockchains/solana/default.nix:
```diff
@@ -0,0 +1,60 @@
+  # on darwin with missing framework System. This framework is not available in nixpkgs
+  # on linux with some librocksdb-sys compilation error
+  doCheck = false;
+  checkInputs = [ pkg-config rustfmt ];
```
This and all of the settings below can be removed now that `doCheck = false`, right? I think it’s better to omit them for now since they don’t do anything, than to keep them where they may confuse others about their purpose.

happysalada replied:
I've commented out stuff only needed for the checkPhase. I would like to keep it for anybody who tries to enable the checkphase.
I've removed the uneeded arguments from inputs though.

ruuda reviewed (commented):
With one more change it now compiles, and I can confirm that the binary works then.

happysalada committed:
solana-testnet: init at 1.9.2

happysalada commented:
I think I've taken care of all your comments. When you have a moment, could you review one more time ?

ruuda reviewed (approved):
LGTM! I can also confirm that I can build this, and the resulting binary works.
